### ISSUE TITLE
Performance of Magick bar code reader

### ISSUE THREAD

**Issue by jackbond** [micjahn/ZXing.Net#98]
The Magick binding for reading bar codes is a lot slower than the System.Drawing implementation. I took a look at the code and was able to improve performance considerably with the following changes to  MagickImageLuminanceSource. Thanks for the great library!

```
private void CalculateLuminance(MagickImage src)
{
    if (src == null)
        throw new ArgumentNullException(nameof(src));

    if (src.HasAlpha)
    {
        CalculateLuminanceBGRA32(src);
    }

    else
    {
        CalculateLuminanceBGR24(src);
    }
}

private void CalculateLuminanceBGR24(MagickImage image)
{
    byte[] imageBytes = image.ToByteArray(MagickFormat.Rgb);

    var luminanceIndex = 0;

    int index = 0;

    for (var y = 0; y < image.Height; y++)
    {
        for (var x = 0; x < image.Width; x++, index += 3)
        {
            byte red = imageBytes[index];
            byte green = imageBytes[index + 1];
            byte blue = imageBytes[index + 2];

            luminances[luminanceIndex] = (byte)((RChannelWeight * red + GChannelWeight * green + BChannelWeight * blue) >> ChannelWeight);
            luminanceIndex++;
        }
    }
}

private void CalculateLuminanceBGRA32(MagickImage image)
{
    byte[] imageBytes = image.ToByteArray(MagickFormat.Rgba);

    var luminanceIndex = 0;

    int index = 0;

    for (var y = 0; y < image.Height; y++)
    {
        for (var x = 0; x < image.Width; x++, index += 4)
        {
            byte red = imageBytes[index];
            byte green = imageBytes[index + 1];
            byte blue = imageBytes[index + 2];
            byte alpha = imageBytes[index + 3];

            var luminance = (byte)((RChannelWeight * red + GChannelWeight * green + BChannelWeight * blue) >> ChannelWeight);
            luminances[luminanceIndex] = (byte)(((luminance * alpha) >> 8) + (255 * (255 - alpha) >> 8));
            luminanceIndex++;
        }
    }
}
```

**Comment by jackbond** [comment]
I can say with near 100% certainty that you've forgotten more about this stuff than I'll ever know. :) The primary issue that I noticed with the original version was that it was calling a lot of per pixel methods, i.e. pixels[x, y].ToColor(), which was triggering a lot of memory allocations and cleanup. If your version works better and faster, that's truly all I care about. Thanks again! I actually did some other pretty good optimizations over the weekend, should I just keep posting them as issues?

**Comment by micjahn** [comment]
The pixel method ToColor has one benefit: I don't have to care about the underlying pixel format of the image. And with GetPixels() I don't have to copy around additional arrays. But the fastest method I'v found yesterday is ToByteArray with a grayscale output. The only thing I don't know is if there are any drawbacks if the BitDepth is bigger than 8.
For your other optimizations it would be great if you can provide them as a pull request. But it is also possible to open additional issues.

**Comment by micjahn** [comment]
Last changes are available in the current version 0.16.5 of the binding package.
